include/matx/kernels/resample_poly.cuh

```diff
@@ -0,0 +1,209 @@
+
+    constexpr int Rank = OutType::Rank();
+
+    // We assume odd-length filters below. In the case of an even-length filter,
+    // logically prepend the filter with a single zero to make its length odd.
+    const bool is_even = filter_len % 2 == 0;
+    if (filter_len % 2 == 0) {
+        filter_len++;
+    }
+
+    const int phase_ind = blockIdx.y;
+    const int tid = threadIdx.x;
+    const int filter_len_half = filter_len/2;
+
+    // All but the last dim are batch indices
+    const int batch_idx = blockIdx.x;
+    auto bdims = BlockToIdx(output, batch_idx, 1);
+
+    // We assume odd-length filters in the below index calculations. If the filter
+    // is even-length, then it has had a single zero logically prepended to it to
+    // make it odd-length.
+    // Consider resampling an input x by up=3, down=2. The upsampled and then
+    // downsampled sequence -- ignoring filtering for now -- will look as folows.
+    // | x0 |  0 |  0 | x1 |  0 |  0 | x2 | ... ]
+    //
+    //   d0        d1        d2        d3   ...
+    // The dX are samples that will be kept after filtering and downsampling.
+    // The filter is centered on the output sample. We want to avoid storing the
+    // filter coefficents that will map to 0 in the upsampled sequence. Thus,
+    // we see in the above that for filter h with length K and output d0, we will
+    // need filter elements h[K/2], h[K/2 - 3], h[K/2 - 6], etc., depending upon
+    // the length of the filter. The filter is flipped for convolution, which is
+    // why the index is decreasing in the above. We call this set of coefficients
+    // phase 0 because there is an offset of 0 between the xN points in the
+    // upsampled sequence and the central tap of the filter.
+    // Similarly, output d1 will use phase 1 coefficients because x1 is offset by
+    // 1 from the central tap, and output d1 will use phase 2 coefficients because
+    // x2 is offset by 2 from the central tap. Thus, there are up total phases and
+    // therefore either filter_len / up or filter_len / up - 1 coefficients per phase.
+    //
+    // This kernel is launched with the phase index provided as a block index
+    // and it will calculate the output points corresponding to this phase. We first
+    // need to determine which filter samples we need for this phase and load them
+    // into shared memory.
+    // left_filter_ind is the filter index that corresponds to the input sample
+    // in the upsampled sequence equal to or to the left of the output sample.
+    // Thus, for phase 0, left_filter_ind is the central filter tap, h[k/2].
+    // For phase 1, with our up=3, down=2 example, left_filter_ind is
+    // h[K/2+2]. Once we have a single tap index and the corresponding x index,
+    // we will stride in both sequences (input and filter) by +/- up to cover
+    // the full filter phase.
+    const int left_filter_ind = filter_len_half + (phase_ind*down) % up;

```

Comment:
Does this need to be `index_t`?

Reply:
I can switch to all index_t in the kernel for now. It will lead to some 64-bit integer divisions in the case of 64-bit index_t, but future optimizations can determine when it can safely be downgraded to 32-bit (e.g. for smem indices).

Reply:
You can compile with `-DMATX_BUILD_32_BIT` it will change `index_t` to a signed 32b int. We have one project internally that relies on this working. You can also specify custom sizes, but that's not tested very well.